nemoguardrails/integrations/langchain/runnable_rails.py

```diff
@@ -0,0 +1,220 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from typing import Any, List, Optional
+
+from langchain_core.language_models import BaseLanguageModel
+from langchain_core.messages import AIMessage, HumanMessage
+from langchain_core.prompt_values import ChatPromptValue, StringPromptValue
+from langchain_core.runnables import Runnable
+from langchain_core.runnables.config import RunnableConfig
+from langchain_core.runnables.utils import Input, Output
+from langchain_core.tools import Tool
+
+from nemoguardrails import LLMRails, RailsConfig
+
+
+class RunnableRails(Runnable[Input, Output]):
+    def __init__(
+        self,
+        config: RailsConfig,
+        llm: Optional[BaseLanguageModel] = None,
+        tools: Optional[List[Tool]] = None,
+        passthrough: bool = True,
+        passthrough_runnable: Optional[Runnable] = None,
+        input_key: str = "input",
+        output_key: str = "output",
+    ) -> None:
+        self.llm = llm
+        self.passthrough = passthrough
+        self.passthrough_runnable = passthrough_runnable
+        self.passthrough_user_input_key = input_key
+        self.passthrough_bot_output_key = output_key
+
+        # We override the config passthrough.
+        config.passthrough = passthrough
+
+        self.rails = LLMRails(config=config, llm=llm)
+
+        if tools:
+            # When tools are used, we disable the passthrough mode.
+            self.passthrough = False
+
+            for tool in tools:
+                self.rails.register_action(tool, tool.name)
+
+        # If we have a passthrough Runnable, we need to register a passthrough fn
+        # that will call it
+        if self.passthrough_runnable:
+            self._init_passthrough_fn()
+
+    def _init_passthrough_fn(self):
+        """Initialize the passthrough function for the LLM rails instance."""
+
+        async def passthrough_fn(context: dict, events: List[dict]):
+            # First, we fetch the input from the context
+            _input = context.get("passthrough_input")
+            _output = await self.passthrough_runnable.ainvoke(input=_input)
+
+            # If the output is a string, we consider it to be the output text
+            if isinstance(_output, str):
+                text = _output
+            else:
+                text = _output.get(self.passthrough_bot_output_key)
+
+            return text, _output
+
+        self.rails.llm_generation_actions.passthrough_fn = passthrough_fn
+
+    def __or__(self, other):
+        if isinstance(other, BaseLanguageModel):
+            self.llm = other
+            self.rails.update_llm(other)
+
+        elif isinstance(other, Runnable):
+            self.passthrough_runnable = other
+            self.passthrough = True
+            self._init_passthrough_fn()
+
+        return self
```

Comment:
Ah I see - this is why you need to wrap them in parentheses. This is not recommended. Instead, I would recommend passing in the `llm` as an parameter to `RunnableRails`

Reply:
That also works. But I figured that supporting the "|" notation would make it easier to go from "without guardrails" to "with guardrails", by following a simple wrapping pattern.